### PR TITLE
8175874: Update Security.insertProviderAt to specify behavior when requested position is out of range.

### PR DESCRIPTION
**Pull request by blperez01**
Fixing the bug as stated would cause compatibility issues. E.g. the `addProvider` function always adds new providers at position 0 and therefore would always throw an error. Instead of changing how the function handles indices, we instead have added this behavior to the function spec.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Change requires CSR request [JDK-8314168](https://bugs.openjdk.org/browse/JDK-8314168) to be approved
- [x] Commit message must refer to an issue

### Issues
 * [JDK-8175874](https://bugs.openjdk.org/browse/JDK-8175874): Update Security.insertProviderAt to specify behavior when requested position is out of range. (**Enhancement** - P4)
 * [JDK-8314168](https://bugs.openjdk.org/browse/JDK-8314168): Update Security.insertProviderAt to specify behavior when requested position is out of range. (**CSR**)

### Reviewers
 * [Sean Mullan](https://openjdk.org/census#mullan) (@seanjmullan - **Reviewer**)
 * [Valerie Peng](https://openjdk.org/census#valeriep) (@valeriepeng - **Reviewer**)
 * @ASISBusiness (no known openjdk.org user name / role)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14925/head:pull/14925` \
`$ git checkout pull/14925`

Update a local copy of the PR: \
`$ git checkout pull/14925` \
`$ git pull https://git.openjdk.org/jdk.git pull/14925/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14925`

View PR using the GUI difftool: \
`$ git pr show -t 14925`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14925.diff">https://git.openjdk.org/jdk/pull/14925.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14925#issuecomment-1640925035)

### CODE REVIEW

**Comment by seanjmullan** [comment]
Minor comment - before integrating, update the PR description to reflect the latest change (it still says we will add an `@implNote`.)

**Comment by seanjmullan** [comment]
I looked around, and I don't think there is an existing test that actually tests this behavior. So, I think we should add one. I think you can probably extend the existing `test/jdk/java/security/Provider/ChangeProviders.java` regression test and add more tests that try to insert the provider at various positions outside of the normal range (ex: -1, 0, n +1, n + 10) , and then checks that the provider was inserted at the end of the list.

File: src/java.base/share/classes/java/security/Security.java [hunk 1]
```diff
@@ -296,7 +296,10 @@ public static String getAlgorithmProperty(String algName,
      * Adds a new provider, at a specified position. The position is
      * the preference order in which providers are searched for
      * requested algorithms.  The position is 1-based, that is,
-     * 1 is most preferred, followed by 2, and so on.
+     * 1 is most preferred, followed by 2, and so on.  If the position
+     * is less than 1 or greater than n, where n is the number of providers,
```
Comment (valeriepeng) [inline src/java.base/share/classes/java/security/Security.java:300]: nit: where n is the number of _installed_ providers

**Comment by valeriepeng** [comment]
> I looked around, and I don't think there is an existing test that actually tests this behavior. So, I think we should add one. I think you can probably extend the existing `test/jdk/java/security/Provider/ChangeProviders.java` regression test and add more tests that try to insert the provider at various positions outside of the normal range (ex: -1, 0, n +1, n + 10) , and then checks that the provider was inserted at the end of the list.

Is a CSR needed? Now that we are enhancing existing javadoc for covering corner case scenarios?

**Comment by seanjmullan** [comment]
> Is a CSR needed? Now that we are enhancing existing javadoc for covering corner case scenarios?

Yes, since the proposed wording will become part of the specification, a CSR is required. CSR has already been filed and approved: https://bugs.openjdk.org/browse/JDK-8314168

File: src/java.base/share/classes/java/security/Security.java [hunk 2]
```diff
@@ -296,7 +296,10 @@ public static String getAlgorithmProperty(String algName,
      * Adds a new provider, at a specified position. The position is
      * the preference order in which providers are searched for
      * requested algorithms.  The position is 1-based, that is,
-     * 1 is most preferred, followed by 2, and so on.
+     * 1 is most preferred, followed by 2, and so on.  If the position
+     * is less than 1 or greater than n, where n is the number of installed
+     * providers, the provider (if not already installed) is inserted at
+     * the end ofthe list, or at the n + 1 position.
```
Comment (seanjmullan) [inline src/java.base/share/classes/java/security/Security.java:302]: Missing space between "of" and "the".

File: test/jdk/java/security/Provider/ChangeProviders.java
Comment (seanjmullan) [inline test/jdk/java/security/Provider/ChangeProviders.java:1]: Add bugid to the `@bug` line.

File: test/jdk/java/security/Provider/ChangeProviders.java [hunk 3]
```diff
@@ -81,6 +81,22 @@ public static void main0(String[] args) throws Exception {
             throw new Exception("Provider not at pos 1");
         }
 
+        //Ensure that providers inserted at position 0 or > n are placed
+        //at the n+1st position
```
Comment (seanjmullan) [inline test/jdk/java/security/Provider/ChangeProviders.java:85]: Nit: typically there is a space after `//`.

File: test/jdk/java/security/Provider/ChangeProviders.java [hunk 4]
```diff
@@ -81,6 +81,22 @@ public static void main0(String[] args) throws Exception {
...
+        class NextProvider extends Provider {
+            public NextProvider() { super("Bar", "47.23", "none"); }
+        }
+        Provider r = new NextProvider();
+        Security.insertProviderAt(r,0);
```
Comment (seanjmullan) [inline test/jdk/java/security/Provider/ChangeProviders.java:90]: Nit: add space after `,`.

File: test/jdk/java/security/Provider/ChangeProviders.java [hunk 5]
```diff
@@ -81,6 +81,22 @@ public static void main0(String[] args) throws Exception {
...
+        if (plen() != n+2 || Security.getProviders()[n+1] != r) {
```
Comment (seanjmullan) [inline test/jdk/java/security/Provider/ChangeProviders.java:91]: Nit: add space before/after `+`.

**Review by seanjmullan — CHANGES_REQUESTED** [review]

File: test/jdk/java/security/Provider/ChangeProviders.java (hunk 3 above)
Comment (seanjmullan) [inline test/jdk/java/security/Provider/ChangeProviders.java:85]: It's probably clearer to reset n after this and then test for n+1 instead of n+2, ex - add a line:

```
// reset n to number of installed providers
n = plen();
```

File: test/jdk/java/security/Provider/ChangeProviders.java [hunk 6]
```diff
@@ -81,6 +81,22 @@ public static void main0(String[] args) throws Exception {
             throw new Exception("Provider not at pos 1");
         }
 
+        //Ensure that providers inserted at position 0 or > n are placed
```
Comment (valeriepeng) [inline test/jdk/java/security/Provider/ChangeProviders.java:84]: nit: this sentence is a bit hard to parse, maybe it's easier to just state that positions out of the [1...n] range is placed at the position n+1?

Comment (valeriepeng) [reply test/jdk/java/security/Provider/ChangeProviders.java:84]: Actually, you don't need another provider impl, you could have just removed Foo and then proceed with the Foo provider with your insertProviderAt(...) testing. With the current test, when you start testing, the provider count is n+1, which can be confusing mixed with your comment of using n as the current provider count... Thus, I'd suggest you just remove the Foo provider and re-use it. This way, there is less code and all is consistent.

File: test/jdk/java/security/Provider/ChangeProviders.java
Comment (valeriepeng) [reply test/jdk/java/security/Provider/ChangeProviders.java:1]: +1, append 8175874 to line 26.
Have no further comments, thanks

**Review by seanjmullan — APPROVED** [review]

**Review by valeriepeng — APPROVED** [review]

**Comment by ASISBusiness** [comment]
> Fixing the bug as stated would cause compatibility issues. E.g. the `addProvider` function always adds new providers at position 0 and therefore would always throw an error. Instead of changing how the function handles indices, we instead have added this behavior to the function spec.
> 
> 
> ---------
> ### Progress
> - [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
> - [x] Change must not contain extraneous whitespace
> - [x] Change requires CSR request [JDK-8314168](https://bugs.openjdk.org/browse/JDK-8314168) to be approved
> - [x] Commit message must refer to an issue
> 
> ### Issues
>  * [JDK-8175874](https://bugs.openjdk.org/browse/JDK-8175874): Update Security.insertProviderAt to specify behavior when requested position is out of range. (**Enhancement** - P4)
>  * [JDK-8314168](https://bugs.openjdk.org/browse/JDK-8314168): Update Security.insertProviderAt to specify behavior when requested position is out of range. (**CSR**)
> 
> 
> ### Reviewers
>  * [Sean Mullan](https://openjdk.org/census#mullan) (@seanjmullan - **Reviewer**)
>  * [Valerie Peng](https://openjdk.org/census#valeriep) (@valeriepeng - **Reviewer**)
> 
> 
> ### Reviewing
> <details><summary>Using <code>git</code></summary>
> 
> Checkout this PR locally: \
> `$ git fetch https://git.openjdk.org/jdk.git pull/14925/head:pull/14925` \
> `$ git checkout pull/14925`
> 
> Update a local copy of the PR: \
> `$ git checkout pull/14925` \
> `$ git pull https://git.openjdk.org/jdk.git pull/14925/head`
> 
> </details>
> <details><summary>Using Skara CLI tools</summary>
> 
> Checkout this PR locally: \
> `$ git pr checkout 14925`
> 
> View PR using the GUI difftool: \
> `$ git pr show -t 14925`
> 
> </details>
> <details><summary>Using diff file</summary>
> 
> Download this PR as a diff file: \
> <a href="https://git.openjdk.org/jdk/pull/14925.diff">https://git.openjdk.org/jdk/pull/14925.diff</a>
> 
> </details>
> 
> 
> ### Webrev
> [Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14925#issuecomment-1640925035)

**Review by ASISBusiness — APPROVED** [review]
Hi @ASISBusiness, thanks for making a comment in an OpenJDK project!

All comments and discussions in the OpenJDK Community must be made available under the OpenJDK [Terms of Use](https://openjdk.java.net/legal/tou/). If you already are an OpenJDK [Author](https://openjdk.java.net/bylaws#author), [Committer](https://openjdk.java.net/bylaws#committer) or [Reviewer](https://openjdk.java.net/bylaws#reviewer), please click [here](https://bugs.openjdk.java.net/secure/CreateIssue.jspa?pid=11300&issuetype=1) to open a new issue so that we can record that fact. Please Use "Add GitHub user ASISBusiness for the summary.

If you are not an OpenJDK Author, Committer or Reviewer, simply check the box below to accept the OpenJDK Terms of Use for your comments.

 - [ ]  I agree to the [OpenJDK Terms of Use](https://openjdk.java.net/legal/tou/) for all comments I make in a project in the [OpenJDK GitHub organization](https://github.com/openjdk).

Your comment will be automatically restored once you have accepted the OpenJDK [Terms of Use](https://openjdk.java.net/legal/tou/).
